examples/virtual-dom-wip/platform/Html/Internal/Client.roc

```diff
@@ -281,9 +266,51 @@ diff = \{ rendered, patches }, newNode ->
             else
                 { rendered, patches }
 
-        { oldNode: RenderedElement _oldName _oldAttrs _oldChildren, newNode: Element _newName _newSize _newAttrs _newChildren } ->
-            # TODO: actual diffing! LOL This is just to get something up and running
-            replaceNode { rendered, patches } rendered.root newNode
+        { oldNode: RenderedElement oldName oldAttrs oldChildren, newNode: Element newName _ newAttrs newChildren } ->
+            if newName != oldName then
+                replaceNode { rendered, patches } root newNode
+            else
+                stateAttrs =
+                    diffAttrs { rendered, patches } root oldAttrs newAttrs
+                stateChildPairs =
+                    List.map2 oldChildren newChildren (\oldChildId, newChild -> { oldChildId, newChild })
+                    |> List.walk stateAttrs \childWalkState, { oldChildId, newChild } ->
+                        { rendered: childWalkRendered, patches: childWalkPatches } = childWalkState
+                        diff { rendered: { childWalkRendered & root: oldChildId }, patches: childWalkPatches } newChild
+                { rendered: renderedLeftOverChildren, patches: patchesLeftOverChildren } =
+                    if List.len oldChildren > List.len newChildren then
+                        List.walkFrom oldChildren (List.len newChildren) stateChildPairs deleteNode
+                    else if List.len oldChildren < List.len newChildren then
+                        stateBeforeCreate = {
+                            rendered: stateChildPairs.rendered,
+                            patches: stateChildPairs.patches,
+                            ids: [],
+                        }
+                        { rendered: renderedAfterCreate, patches: patchesAfterCreate, ids: createdIds } =
+                            List.walkFrom newChildren (List.len oldChildren) stateBeforeCreate createChildNode
+                        # Find the children again since they might have new node IDs!
+                        latestNode =
+                            List.get renderedAfterCreate.nodes root
+                            |> Result.withDefault (Ok RenderedNone)
+                            |> Result.withDefault (RenderedNone)
+                        nodeWithUpdatedChildren =
+                            when latestNode is
+                                RenderedElement n a c -> RenderedElement n a (List.concat c createdIds)
+                                _ -> latestNode # impossible
```

Comment:
Maybe a good place to use `crash` here instead of defaulting to `RenderedNone`? Or maybe to add an expect?

Reply:
That's done. I also ended up merging it with the next expression, it's much nicer! Good suggestion.